Sources/WhoopDIKit/Module/AccumulationDefinition.swift

```diff
@@ -0,0 +1,121 @@
+        self.serviceKey = ServiceKey(Key.FinalValue.self, name: name)
+        self.valueProvider = wrapAccumulationProvider(name: name, accumulationKey: accumulationKey, valueProvider: valueProvider)
+        self.insertionFunc = { dictionary, definition in accumulate(key: accumulationKey,
+                                                                    definition: definition,
+                                                                    accumulatedValueGetter: valueProvider,
+                                                                    serviceDictionary: dictionary)
+        }
+    }
+
+    func get(params: Any?, container: Container) throws -> Any {
+        try valueProvider(params, container)
+    }
+
+    func insert(into serviceDictionary: ServiceDictionary<any DependencyDefinition>) {
+        insertionFunc(serviceDictionary, self)
+    }
+}
+
+/// Provides the definition of an accumulation that is calculated once and cached.
+/// The accumulated value is computed once per container and reused on subsequent requests.
+final class SingletonAccumulationDefinition: DependencyDefinition {
+    private let accumulationKeyType: Any.Type
+    private let valueProvider: (Any?, Container) throws -> Any
+    private let lock = NSLock()
+    let serviceKey: ServiceKey
+    private let insertionFunc: (ServiceDictionary<DependencyDefinition>, SingletonAccumulationDefinition) -> Void
+    private var cachedAccumulated: Any? = nil
+    private var cachedValue: Any? = nil
+
+    init<Key: AccumulationKey>(name: String?, accumulationKey: Key.Type, valueProvider: @escaping (Any?) throws -> Key.AccumulatedValue) {
+        self.accumulationKeyType = accumulationKey
+        self.serviceKey = ServiceKey(Key.FinalValue.self, name: name)
+        self.valueProvider = wrapAccumulationProvider(name: name, accumulationKey: accumulationKey, valueProvider: valueProvider)
+        self.insertionFunc = { dictionary, definition in accumulate(key: accumulationKey,
+                                                                    definition: definition,
+                                                                    accumulatedValueGetter: { params in
+            try definition.lockValue(externalGet: {
+                try valueProvider(params)
+            }, cache: \.cachedAccumulated)
+        },
+                                                                    serviceDictionary: dictionary)
+        }
+    }
+
+    private func lockValue<T>(externalGet: () throws -> T, cache: ReferenceWritableKeyPath<SingletonAccumulationDefinition, Any?>) throws -> T {
+        if let value = self[keyPath: cache] {
+            // Double-checked locking for the accumulated value
+            return value as! T
+        }
+
+        lock.lock()
+        defer { lock.unlock() }
+
+        if let value = self[keyPath: cache] {
+            return value as! T
+        }
+
+        let value = try externalGet()
+
+        self[keyPath: cache] = value
```

Comment:
This is kind of confusing:
1. Why do we need to use a key path here? 
2. Why are we storing two different values in two different variables?
3. The threading implications of double check locking across two variables, with a less atomic read/write is more challenging to think through. It's not clear to me if there's potential races here.